### PR TITLE
feat: change BinFinder logic

### PR DESCRIPTION
**Pull request by LuisFelipeCoelho**
This PR changes the `BinFinder` algorithm that returns the connected bins used to build seeds. Instead of always returning the 8 surrounding bins, the user may define a vector containing a map of z bins in the top and bottom sets of space points for each middle space point bin. The algorithm will only search for SP in the bins defined in the map. In case no map is defined, the algorithm returns the 8 surrounding bins for top and bottom layers.
That will allow us to configure the top and bottom maps independently form each other and also to
return the bins based on geometrical assumptions as it is done by the ATLAS algorithm, for example:

```
//  1     2     3     4     5     6     7     8     9     10     11     z bin index
// -------------------------------------------------------------------> Z[mm]
// vector containing the map of z bins in the top layer
seedingCfg.binFinderConfigTop = {{1},{1, 2},{2, 3},{3, 4},{4, 5},{5, 6, 7},{7, 8},{8, 9},{9, 10},{10, 11},{11}};
// vector containing the map of z bins in the bottom layer
seedingCfg.binFinderConfigBottom = {{1, 2},{2, 3},{3, 4},{4, 5},{5, 6},{6},{6, 7},{7, 8},{8, 9},{9, 10},{10, 11}};
```

these vectors are passed to `BinFinder` by the `SeedingAlgorithm`.

@noemina

### CODE REVIEW

**Comment by paulgessinger** [comment]
@robertlangenberg can we get this in?

**Comment by robertlangenberg** [comment]
> @robertlangenberg can we get this in?

@paulgessinger this PR depends on 1052, so it should only go in after 1052. Your review is requested ;)

**Comment by robertlangenberg** [comment]
Hey @LuisFelipeCoelho , is this still WIP now that #1052 went in?

**Comment by LuisFelipeCoelho** [comment]
No, I think it isn't. @robertlangenberg

**Review by robertlangenberg — APPROVED** [review]
looks good, let's wait for the pending CI checks

File: Core/include/Acts/Seeding/SeedfinderConfig.hpp
Comment (robertlangenberg) [inline Core/include/Acts/Seeding/SeedfinderConfig.hpp:82]: This doesn't need to be here, right? Is this used anywhere? Since this is not part of the Seedfinder itself but rather the grid and the Binfinder, I'd leave the config confined to there.

**Review by robertlangenberg — CHANGES_REQUESTED** [review]

**Comment by LuisFelipeCoelho** [comment]
@robertlangenberg I added that because I needed to change the `bottomBinFinder` and `topBinFinder` in `SeedfinderTest.cpp ` , but I am not sure if this is the best way of doing that

File: Tests/UnitTests/Core/Seeding/SeedfinderTest.cpp
Comment (robertlangenberg) [inline Tests/UnitTests/Core/Seeding/SeedfinderTest.cpp:144]: please don't move individual parts out of a config that you continue to use later. By std::move-ing them, they members of the config struct become invalid.

as mentioned above, you don't need to have the neighbor bins in the SeedfinderConfig at all, you can just create them separately.

File: Examples/Algorithms/TrackFinding/src/SeedingAlgorithm.cpp
Comment (robertlangenberg) [inline Examples/Algorithms/TrackFinding/src/SeedingAlgorithm.cpp:131]: please don't move individual parts out of a config that you continue to use later. By std::move-ing them, they members of the config struct become invalid.

as mentioned above, you don't need to have the neighbor bins in the SeedfinderConfig at all, you can just create them separately.

File: Core/include/Acts/Seeding/SeedfinderConfig.hpp
Comment (robertlangenberg) [reply Core/include/Acts/Seeding/SeedfinderConfig.hpp:82]: Why do you want to have them in any configuration? For the python tests?

File: Examples/Algorithms/TrackFinding/src/SeedingAlgorithm.cpp
Comment (robertlangenberg) [reply Examples/Algorithms/TrackFinding/src/SeedingAlgorithm.cpp:131]: please only mark a conversation as resolved after changing it in the code, makes reviewing easier :)

**Review by robertlangenberg — APPROVED** [review]
Changes look good!

**Review by paulgessinger — APPROVED** [review]
Minimal changes after @robertlangenberg's recent approval, reapproving to get this in.

**Comment by paulgessinger** [comment]
@LuisFelipeCoelho the SYCL build seems to choke:

```console
FAILED: Tests/UnitTests/Plugins/Sycl/Seeding/CMakeFiles/ActsUnitTestSeedfinderSycl.dir/SeedfinderSyclTest.cpp.o 
/opt/intel/oneapi/compiler/2021.1.1/linux/bin/clang++  -DBOOST_ALL_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DVECMEM_DEBUG_MSG_LVL=0 -DVECMEM_HAVE_EXPERIMENTAL_PMR_MEMORY_RESOURCE -DVECMEM_SOURCE_DIR_LENGTH=15 -I../Core/include -ICore -I../Plugins/Sycl/include -I_deps/vecmem-src/core/include -I_deps/vecmem-src/sycl/include -isystem /usr/include/eigen3 -Werror -Wall -Wextra -Wpedantic -Wshadow -Wunused-local-typedefs -O3 -DNDEBUG    -std=gnu++17 -MD -MT Tests/UnitTests/Plugins/Sycl/Seeding/CMakeFiles/ActsUnitTestSeedfinderSycl.dir/SeedfinderSyclTest.cpp.o -MF Tests/UnitTests/Plugins/Sycl/Seeding/CMakeFiles/ActsUnitTestSeedfinderSycl.dir/SeedfinderSyclTest.cpp.o.d -o Tests/UnitTests/Plugins/Sycl/Seeding/CMakeFiles/ActsUnitTestSeedfinderSycl.dir/SeedfinderSyclTest.cpp.o -c ../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp
../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:106:7: error: unused variable 'numPhiNeighbors' [-Werror,-Wunused-variable]
  int numPhiNeighbors = 1;
      ^
../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:146:35: error: use of undeclared identifier 'zBinNeighborsBottom'
      Acts::BinFinder<SpacePoint>(zBinNeighborsBottom, numPhiNeighbors));
                                  ^
../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:146:56: error: use of undeclared identifier 'numPhiNeighbors'
      Acts::BinFinder<SpacePoint>(zBinNeighborsBottom, numPhiNeighbors));
                                                       ^
../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:148:35: error: use of undeclared identifier 'zBinNeighborsTop'
      Acts::BinFinder<SpacePoint>(zBinNeighborsTop, numPhiNeighbors));
                                  ^
../Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:148:53: error: use of undeclared identifier 'numPhiNeighbors'
      Acts::BinFinder<SpacePoint>(zBinNeighborsTop, numPhiNeighbors));
                                                    ^
5 errors generated.
```

**Review by paulgessinger — APPROVED** [review]

**Comment by paulgessinger** [comment]
I now realize this was actually a breaking change. (Just commenting here for reference)

**Pull request merged**
